### PR TITLE
Allow templated strings instead of options for customizing verse callout format

### PR DESCRIPTION
**Pull request by brennenputh**
With all the customization that needs to be provided to allow users the freedom to format how they want, the configuration options will quickly explode in size.  This PR instead implements a template system for the header and footer, allowing much greater flexibility for less configuration options.

Pain points are mainly related to moving from the old system to the new one, but are not extensive.  A problem that is difficult to address is whether the collapsible verses option should still be around or be replaced by this system.  A hackfix is implemented here, but it is not a great fix.

Unfinished aspects of this PR:
- [x] UI for presenting templates to the user
- [x] Automatic conversion from the old configuration options to the templates

### CODE REVIEW

**Comment by brennenputh** [comment]
I have finished the migration code, however it requires more testing than I have.  Consider this a call for testers, this is ready for anyone who's willing.

**Comment by ajrty33** [comment]
> I have finished the migration code. Consider this a call for testers, this is ready for anyone who's willing.

I'll test it out!
Until then, it looks like `> [!Bible]` is left out of the customization? If that is added as a header option, then it can be removed to allow a minimal output that just shows up like a quote like:
> Verse text
> Verse reference

**Comment by brennenputh** [comment]
I considered adding it to the format string and not appending, but was unsure.  If that's a desired feature, it's not hard to make configurable.

**Comment by brennenputh** [comment]
The callout prefixes are now in the customization - both for the header and the footer.  Also, added a nice bright red warning on the migrator.

**Comment by RandomUserWithInternet** [comment]
What is the purpose of `> [!Bible]`?

**Comment by brennenputh** [comment]
> What is the purpose of `> [!Bible]`? 

https://help.obsidian.md/Editing+and+formatting/Callouts

**Comment by tim-hub** [comment]
Thanks for all of this, I will try to give it a test too by weekends.

**Comment by brennenputh** [comment]
New task that needs completed:
- [ ] Reload settings page after migration

I'm moving atm, so I may not be able to do much for a week or two.  I don't believe it is the largest problem to block merge, but certainly something that should be done.

File: src/ui/BibleReferenceSettingTab.ts
Comment (tim-hub) [inline src/ui/BibleReferenceSettingTab.ts:199]: to clarify, this is a broken change, right?

Comment (tim-hub) [reply src/ui/BibleReferenceSettingTab.ts:199]: and after migration, both old settings, and new settings ui existed in the setting page.

Is it?

Comment (tim-hub) [reply src/ui/BibleReferenceSettingTab.ts:199]: I personally feel this is a bit confusing? of this migration warning.

image

If it is a safe migration, I guess it is OK to just migrate it by default, and fully update the setting UI page, just remove the old ones. 

`> [!Bible] {{verse_reference_link}}` I personally not prefer in this way (template) for header and footer separately


If we want to have a template I would rather have an `Advanced UI` section to have a template for the whole rendering in text area, something like 


```
> [!Bible] {{verse_reference_link}}

{{verses}}

{{chapter}}{{chapterNumber}}{{verseStart}}

```

Comment (tim-hub) [reply src/ui/BibleReferenceSettingTab.ts:199]: > I personally feel this is a bit confusing? of this migration warning.
> 
> image
> 
> If it is a safe migration, I guess it is OK to just migrate it by default, and fully update the setting UI page, just remove the old ones.
> 
> `> [!Bible] {{verse_reference_link}}` I personally not prefer in this way (template) for header and footer separately
> 
> If we want to have a template I would rather have an `Advanced UI` section to have a template for the whole rendering in text area, something like
> 
> ```
> > [!Bible] {{verse_reference_link}}
> 
> {{verses}}
> 
> {{chapter}}{{chapterNumber}}{{verseStart}}
> ```
Firstly, thanks so much for the contribution. I have some different thoughts about the UI change.


In sum,  what I want to say is.

IMHO, I will suggest.

- Add an `Advance UI` section for template, instead of two for header, footer
- Use Text Area for template for whole bible verses content
- Not make broken changes, mapping the new template to the old structure of settings, (and remove the broken change warning)

**Review by tim-hub — CHANGES_REQUESTED** [review]
Thanks again for the investigation and work, I left some comments here to explain my thoughts about UI and how to structure the settings

https://github.com/tim-hub/obsidian-bible-reference/pull/109/files#r1299301939

File: src/ui/BibleReferenceSettingTab.ts
Comment (brennenputh) [reply src/ui/BibleReferenceSettingTab.ts:199]: > Add an Advance UI section for template, instead of two for header, footer

I didn't see a _super_ clean way to do this, due to the way verses are formatted.  Definitely something I can look into though.  Formatting verses is a harder thing, as it isn't as simple as `replaceAll` on a string, however I have a few thoughts about the path forward for this (and admittedly, I eventually wanted to get there anyways).

> Use Text Area for template for whole bible verses content

Missed that - I'll get it when I next go over

> Not make broken changes, mapping the new template to the old structure of settings, (and remove the broken change warning)

The age old problem of migration...
Really, what I want to avoid is having two ways to do any one thing.  In my opinion the template is much clearer about what will go where.  Also, there's no easy way to have both.  Either you're formatting by the settings or you're formatting by the template, which just adds another boolean option to the settings. (or the settings update the template, which is somewhat annoying because it would completely overwrite settings if you changed a value)

I believe the middle ground here would be automatically migrating settings into the new system.  This way, no one should notice a break, but everyone can be on the new system, reducing code complexity.  I wasn't entirely sure how to go about this, but I will attempt some things in my next pass.

We still need a way to manually trigger a migration, was it to not automatically trigger for someone.  The migration button could become a command - I believe that would be more sensible for hiding it from the user unless they want it.  It'll still have a warning, but it'll feel more complete and professional in the context of a entire window prompt I believe.

At the end of all this, I do want to say I believe the average user of this plugin should be able to understand the templating syntax, as it is highly similar to the inbuilt templates plugin (on purpose).  Whether this change is a issue for users should be completely up to how breaking it is, and if it will unexpectedly cause people to lose their formatting.

Comment (brennenputh) [reply src/ui/BibleReferenceSettingTab.ts:199]: On second thought, with respect to the possible issue of onboarding users into the templating system, a dialogue could be provided that allows you to build your template string.  Basically just a configurator that provides the most common options (such as the ones provided before) and inserts a template string based on those options into your settings.

**Comment by brennenputh** [comment]
I have added a basic auto migration trigger, however it is quite basic and will run when a user has installed the plugin for the first time.

- [ ] Make the auto migration only trigger when a user is upgrading the plugin, not installing

File: src/VerseSuggesting.ts
Comment (tim-hub) [inline src/VerseSuggesting.ts:42]: I belive the broken change is not necessary, it could be converted from old setting to new template settings without broken it, if change it like this

```suggestion
      if (this.settings?.collapsibleVerses) {
      oldhead += '-'
    }
    if (
      this.settings.referenceLinkPosition ===
        BibleVerseReferenceLinkPosition.Header ||
      this.settings.referenceLinkPosition ===
        BibleVerseReferenceLinkPosition.AllAbove
    ) {
      oldhead += ' {{verse_reference_link}}'
    }
    let head = this.formatReferenceString(this.settings.headFormatString || oldhead)
```

File: src/VerseSuggesting.ts
Comment (tim-hub) [inline src/VerseSuggesting.ts:44]: same here

File: src/VerseSuggesting.ts
Comment (tim-hub) [reply src/VerseSuggesting.ts:42]: This is what I mean in the last comment the settings should be translateble between old and new.


Ref: https://github.com/tim-hub/obsidian-bible-reference/pull/109#discussion_r1299301939

File: src/ui/BibleReferenceSettingTab.ts
Comment (tim-hub) [reply src/ui/BibleReferenceSettingTab.ts]: I agree, I dislike the two ways of doing one thing as well, 

But this new UI contained two ways of doing one thing.




Both the new template and old dropdown can config header and bottom.


This is a big UI change for settings. At least we should not show both template settings, and old header setting in one view  at the same time, it need to be splitted, one need to be hided (through toggle or something) at least, to avoid this UI confusion.

File: src/main.ts
Comment (tim-hub) [inline src/main.ts:48]: I don't believe migration is necessary, settings are translatable vice versa.

File: src/ui/BibleReferenceSettingTab.ts
Comment (tim-hub) [inline src/ui/BibleReferenceSettingTab.ts:196]: If we want to have a template, I believe template should contain 

- header
- content
- footer

For a more industry standard way of template, We should not have a template for header, then a template for footer, so this should be one text area. 

{{[!Bible]}}

{{Content}}

{{Footer}}

**Review by tim-hub — CHANGES_REQUESTED** [review]
Thanks again for contribution and response.


IMHO,

- I like the template idea, however I think if it is a template, then it should be a template not only for header and footer, but for the whole thing.
  - If we want to follow `KISS`, to not allow two UI to update the same settings, I will suggest put template settings UI in `Advanced Settings` and hide it by default until toggle.
- I disagree with broken changes to change settings, for example, as a user who used `syncthing` to sync my obsidian, if one computer update the plugin, the other one does not, the broken change may break the plugin, this is too risky, and I don't think it is necessary

**Comment by tim-hub** [comment]
Again, thanks so much for the contribution, 

Please excuse me, as the project main maintainer, I have to be very careful to approve such a big settings change.

Personally, I like `KISS` too both in codes and UI, codes should be simple for dev, UI should be clear for users.

I left some comments and requests to changes. The Requests to Changes are big too. As I said above, there are some risk above, I have to be careful

File: src/VerseSuggesting.ts
Comment (brennenputh) [reply src/VerseSuggesting.ts:42]: I am unfamiliar with the `string || string` syntax and can't seem to find a reference through Google, is there a link to how that should behave somewhere?

**Comment by brennenputh** [comment]
I am going to remove the additions to the README, as it seems that similar content has moved to the wiki.  For reference, below is the markdown block that was added to the README to explain the formatting options.

```md
### Formatting Strings

These are special substitutions used in the formatting settings.

| Template String          | Substitution                 |
|--------------------------|------------------------------|
| {{bible_version}}        | Current Bible version        |
| {{book}}                 | Name of the quoted book      |
| {{chapter}}              | Number of the quoted chapter |
| {{verse_range}}          | Range of quoted verses       |
| {{verse_reference_link}} | Link to the API reference    |
| {{content}}              | The verse content            |


```

**Comment by brennenputh** [comment]
(note - not ready for re-review yet, I wasn't expecting the merge to be as large as it was)

**Comment by brennenputh** [comment]
I believe the main blocker here is still the old formatting settings and what is happening with them - is this correct?  I don't believe it makes sense to have both types even if one is hidden behind a toggle, as all functions of the older type can be served by the newer with a basic understanding of markdown and templating syntax (and are easy to migrate on startup).
That said, I do understand the desire for simpler settings, and so I am going to try to create a dialog to let users simply configure the template that resembles the older settings page.  Maybe this is a silly idea, but I believe it would be the most straightforward way of simplifying the code.

**Comment by tim-hub** [comment]
> I believe the main blocker here is still the old formatting settings and what is happening with them - is this correct? I don't believe it makes sense to have both types even if one is hidden behind a toggle, as all functions of the older type can be served by the newer with a basic understanding of markdown and templating syntax (and are easy to migrate on startup). That said, I do understand the desire for simpler settings, and so I am going to try to create a dialog to let users simply configure the template that resembles the older settings page. Maybe this is a silly idea, but I believe it would be the most straightforward way of simplifying the code.

In my own opinion,

`translator` / `converter` will be better than migrator.

**Comment by tim-hub** [comment]
> I am going to remove the additions to the README, as it seems that similar content has moved to the wiki. For reference, below is the markdown block that was added to the README to explain the formatting options.
> 
> ```
> ### Formatting Strings
> 
> These are special substitutions used in the formatting settings.
> 
> | Template String          | Substitution                 |
> |--------------------------|------------------------------|
> | {{bible_version}}        | Current Bible version        |
> | {{book}}                 | Name of the quoted book      |
> | {{chapter}}              | Number of the quoted chapter |
> | {{verse_range}}          | Range of quoted verses       |
> | {{verse_reference_link}} | Link to the API reference    |
> | {{content}}              | The verse content            |
> ```

If possible, do you mind taking a screenshot for the UI change?

File: src/VerseSuggesting.ts
Comment (tim-hub) [inline src/VerseSuggesting.ts:1]: Thanks again, 

I think this PR need a rebase, 

`src/VerseSuggesting.ts` was refactored to be under a folder.

**Review by tim-hub — CHANGES_REQUESTED** [review]

**Comment by tim-hub** [comment]
> I am going to remove the additions to the README, as it seems that similar content has moved to the wiki. For reference, below is the markdown block that was added to the README to explain the formatting options.
> 
> ```
> ### Formatting Strings
> 
> These are special substitutions used in the formatting settings.
> 
> | Template String          | Substitution                 |
> |--------------------------|------------------------------|
> | {{bible_version}}        | Current Bible version        |
> | {{book}}                 | Name of the quoted book      |
> | {{chapter}}              | Number of the quoted chapter |
> | {{verse_range}}          | Range of quoted verses       |
> | {{verse_reference_link}} | Link to the API reference    |
> | {{content}}              | The verse content            |
> ```

Maybe something like this.

image

This is from another obsidian plugin

**Pull request closed without merging**
